src/Execution/ModelsLoader/SimpleModelsLoader.php

```diff
@@ -18,15 +19,35 @@ class SimpleModelsLoader implements ModelsLoader
      */
     protected $decorateBuilder;
 
+    /**
+     * Unique key under which to store the relation to eliminate duplicates.
+     *
+     * @var string
+     */
+    protected $key;
+
     public function __construct(string $relation, Closure $decorateBuilder)
     {
         $this->relation = $relation;
         $this->decorateBuilder = $decorateBuilder;
+        $this->key = Str::uuid()->toString();
     }
 
     public function load(EloquentCollection $parents): void
     {
+        // Dot notation may be used to eager load nested relations
+        $parts = explode('.', $this->relation);
+
+        // We just return the first level of relations for now.
+        // They hold the nested relations in case they are needed.
+        $firstRelation = $parts[0];
+
         $parents->load([$this->relation => $this->decorateBuilder]);
+
+        foreach ($parents as $model) {
+            $relation = $model->getRelation($firstRelation);
+            $model->setRelation($this->key, $relation);
+        }
```

Comment:
Not sure I get this loop 🤔 
It seems like we keep setting the same relation as we are doing it based on the key. If this is the desired result, can't we just take the last element in the parents array and set that?

Reply:
Totally rewrote it.